File: Code/Projects/CsTextFinder/Output/Output.cs
using System.Text;
using System.Text.RegularExpressions;

namespace Output;

public class Output
{
    private readonly bool _hide;
    private bool   _dirPrinted;
    private int    _matchCount;
    private string _currentDir = string.Empty;
    private Regex  _regex = new(".");

    public int MatchCount => _matchCount;

    public Output(bool hide = true) => _hide = hide;

    public void SetRegex(string pattern)
    {
        try   { _regex = new Regex(pattern); }
        catch (ArgumentException) { _regex = new Regex("."); }
    }

    public void OnDir(string dirPath)
    {
        _currentDir = dirPath;
        _dirPrinted = false;
        if (!_hide)
        {
            Console.WriteLine($"\n  {_currentDir}");
            _dirPrinted = true;
        }
    }

    public void OnFile(string fileName)
    {
        string fullPath = $"{_currentDir}/{fileName}";
        if (!Find(fullPath)) return;

        if (_hide && !_dirPrinted)
        {
            Console.WriteLine($"\n  {_currentDir}");
            _dirPrinted = true;
        }
        Console.WriteLine($"      {fileName}");
        _matchCount++;
    }

    private bool Find(string filePath)
    {
        string? contents = ReadFile(filePath);
        return contents is not null && _regex.IsMatch(contents);
    }

    private static string? ReadFile(string filePath)
    {
        try { return File.ReadAllText(filePath); }
        catch (UnauthorizedAccessException) { return null; }  // won't succeed on retry
        catch (IOException) { }                               // retry as Latin-1 for binary content
        try { return Encoding.Latin1.GetString(File.ReadAllBytes(filePath)); }
        catch (Exception) { return null; }
    }
}

File: Code/Projects/PyTextFinder/Output/output.py
import re


class Output:
    def __init__(self, hide: bool = True) -> None:
        self._hide = hide
        self._dir_printed = False
        self._match_count = 0
        self._current_dir = ""
        self._regex = re.compile(".")

    @property
    def match_count(self) -> int:
        return self._match_count

    def set_hide(self, hide: bool) -> None:
        self._hide = hide

    def set_regex(self, pattern: str) -> None:
        try:
            self._regex = re.compile(pattern)
        except re.error:
            self._regex = re.compile(".")

    def on_dir(self, dir_path: str) -> None:
        self._current_dir = dir_path
        self._dir_printed = False
        if not self._hide:
            print(f"\n  {self._current_dir}", flush=True)
            self._dir_printed = True

    def on_file(self, file_name: str) -> None:
        full_path = f"{self._current_dir}/{file_name}"
        if not self._find(full_path):
            return
        if self._hide and not self._dir_printed:
            print(f"\n  {self._current_dir}")
            self._dir_printed = True
        print(f"      {file_name}", flush=True)
        self._match_count += 1

    def _read_file(self, file_path: str) -> str | None:
        for encoding in ("utf-8", "latin-1"):
            try:
                with open(file_path, "r", encoding=encoding) as f:
                    return f.read()
            except (UnicodeDecodeError, LookupError):
                continue
            except OSError:
                return None
        return None

    def _find(self, file_path: str) -> bool:
        if self._regex.pattern == ".":
            return True
        contents = self._read_file(file_path)
        return contents is not None and bool(self._regex.search(contents))
